In [6]:
import os
import googleapiclient.discovery
from googleapiclient.errors import HttpError
from googleapiclient.discovery import build
import pandas as pd

In [5]:
YouTube_API_Key = 'AIzaSyC647_bWjOlvBk07WRjzIKyrKXMK3e2Juk'

In [4]:
fox_channel_id = 'UCXIJgqnII2ZOINSWNOGFThA'
abc_channel_id = 'UCBi2mrWuNuyYy4gbM6fU18Q'
msnbc_channel_id = 'UCaXkIU1QidjPwiAYu6GcHjg'

## First Version

In [12]:
def get_video_comments(youtube, video_id, max_results):
    request = youtube.commentThreads().list(
        part="snippet",
        videoId=video_id,
        maxResults=max_results
    )
    response = request.execute()
    
    comments = []

    for item in response["items"]:
        comment = item["snippet"]["topLevelComment"]["snippet"]
        comments.append([
            comment['textDisplay'],
            comment['authorDisplayName'],
            comment['publishedAt'],
            comment['likeCount']
        ])

    df = pd.DataFrame(comments, columns=['comment', 'user', 'publish date', 'like count'])

    return df

In [16]:
youtube = build('youtube', 'v3', developerKey=YouTube_API_Key)

comments_df = get_video_comments(youtube, "K-K2t4axqcg", 1000)

print(comments_df.shape)
comments_df.head()

(100, 4)


comment                user  \
0  First case and death was in Washington state.....           @Lilraw85   
1                       Am scared school close to us   @TuanNguyen-vb8eh   
2  AS MANY AS 300,000 DEATHS IN THE USA BY FALL T...  @legendhasitstudio   
3  I heard that people are self quarantining afte...          @nanah3638   
4  Ginger garlic onions builds up the immune syst...     @bluevelvet5365   

           publish date  like count  
0  2020-03-30T19:00:25Z           0  
1  2020-03-15T14:47:10Z           0  
2  2020-03-02T21:02:52Z           0  
3  2020-03-02T17:50:33Z           0  
4  2020-03-02T13:29:44Z           2

Observation: comments returned are sorted by dates (oldest to newest) when comparing to YouTube video watch page

In [19]:
comments_df = get_video_comments(youtube, "QwIZdQVJAL4", 1000)

print(comments_df.shape)
comments_df.head()

(100, 4)


comment                user  \
0            Kichael Moore is a liberal butt pirate.    @user-jq9oo2tx8i   
1                       Aged like a fine cheese. 🧀 🤣       @joecamel4006   
2     Michael Moore the biggest demtard clown ever .    @davemartino5997   
3                                               LIES           @bac0n406   
4  I&#39;ll take health advice from a 350 lb slob...  @SaundersBenevolus   

           publish date  like count  
0  2023-10-28T05:40:05Z           0  
1  2023-04-20T02:56:27Z           0  
2  2022-11-30T21:44:07Z           2  
3  2022-03-15T01:47:15Z           1  
4  2021-08-22T00:18:29Z           2

Probably  because of the YouTube comment page-turner, can only load 100 comments even set the max to be 1000

## Second Version

In [23]:
def get_video_comments(video_id): 
    youtube = build('youtube', 'v3', developerKey=YouTube_API_Key)
    comments = []
    comments_count = 0
    video_response=youtube.commentThreads().list(
    part='snippet,replies',
    videoId=video_id
    ).execute()
 
    while video_response:
       
        for item in video_response['items']:
            comment = item['snippet']['topLevelComment']['snippet']['textDisplay']
            comments.append(comment)
            comments_count += 1

        if 'nextPageToken' in video_response:
            video_response = youtube.commentThreads().list(
                    part = 'snippet,replies',
                    videoId = video_id,
                      pageToken = video_response['nextPageToken']
                ).execute()
        else:
            break
    print(comments_count)
    return comments
 

In [24]:
video_id = "WaVDZBAxdCs"
comments = get_video_comments(video_id)

363


In [25]:
comments

['8 cases? 1.4 million now',
 'He was told to “self quarantine” now it’s all over MA',
 'If anything this emboldens the need for <a href="http://www.youtube.com/results?search_query=%23MedicareForAll">#MedicareForAll</a> and paid sick leave.. we must do the right thing for generations to come and vote <a href="http://www.youtube.com/results?search_query=%23Bernie2020">#Bernie2020</a> <a href="http://www.youtube.com/results?search_query=%23NotMeUs">#NotMeUs</a>',
 'It was stolen from a canadian lab by Chinese agents',
 'I see that a pattern has emerged! Every time they do a &quot;rescue mission&quot; they are not really rescuing anyone. The plane lands a couple of days later the first cases appear! Then the numbers start to increase! The great repatriation has spread the virus all over the globe! I searched for a corona free country but could only find a handfull of places. WHOS IDEA WAS IT TO SEND ALL THESE PEOPLE TO EVERY CORNER OF THE GLOBE?',
 'Dalai Lama,💩 back to China 🇮🇳🇵🇰',
 'I 

The YouTube video has 802 comments on the YouTube page, but only scraped 363 comments using the YouTube API

In [26]:
video_id = "ZrWAqpPGAxQ"
comments = get_video_comments(video_id)

1974


The YouTube video has 3,226 comments on the YouTube page, but only scraped 1974 comments using the YouTube API

In [27]:
comments

['This has aged like MILK in a dry basement. <a href="about:invalid#zCSafez"></a> Y&#39;all were lied to by a maniac.',
 ',./..',
 'It causes inflammation and serious adverse reactions.',
 '3 years going forward.  The comments are very different.',
 'You should always believe the government 😂  We&#39;re still over in Red China making bio weapons. This network is like the North Korean peoples channel😂',
 'LOL Everybody I know that has taken that vaccine has gotten Covid. You cannot vaccinate anything in the cold and flu family because it changes too quickly.  That’s why the non-vaccinated and the vaccinated infection numbers mirror each other.',
 'Stop requiring it for college students and to keep your job!',
 'Oh boy!  Betcha this video gets deleted cause it&#39;s definitely not aging well!',
 'SINCE 2019,WU HAN,MILITARY SPORTS♥DR. FAUCI and T , appeared on the media many times!  Something,,,,',
 'you get me to be HONEST MAN,♥Before the END, tell the WORLD WHAT WAS GOING ON',
 'This ma

Also the scraped comments are sorted by date from newest to oldest